File: app/services/cache.py
"""Disk-based result cache with a lock-file + TTL pattern.

Mirrors the approach used by proxy.php / goes_tile.py in the hurricanes
site: a `<key>.lock` file marks "rendering in progress" (cleared once
generation finishes or the lock goes stale), and a `<key>.json` file holds
the final result (status: ready|error). FastAPI BackgroundTasks replace the
PHP `nohup` subprocess spawn — everything runs in-process here.
"""
import json
import time
from pathlib import Path
from typing import Optional


class ResultCache:
    def __init__(self, base_dir: Path, lock_timeout: int = 600):
        self.base_dir = base_dir
        self.lock_timeout = lock_timeout
        self.base_dir.mkdir(parents=True, exist_ok=True)

    def _paths(self, key: str) -> tuple[Path, Path]:
        return self.base_dir / f"{key}.json", self.base_dir / f"{key}.lock"

    def get_status(self, key: str) -> Optional[dict]:
        json_path, lock_path = self._paths(key)
        if json_path.exists():
            meta = json.loads(json_path.read_text())
            if meta.get("status") in ("ready", "error"):
                return meta
        if lock_path.exists():
            age = time.time() - lock_path.stat().st_mtime
            if age > self.lock_timeout:
                lock_path.unlink(missing_ok=True)
                return None
            params = {}
            try:
                loaded = json.loads(lock_path.read_text())
                if isinstance(loaded, dict):
                    params = loaded
            except (json.JSONDecodeError, OSError):
                pass  # pre-existing plain-timestamp lock file, or a race with acquire_lock's write
            return {**params, "status": "generating", "key": key, "elapsed": int(age)}
        return None

    def acquire_lock(self, key: str, params: Optional[dict] = None) -> None:
        """`params` (band/cmap/satellite/center/etc.) is whatever the caller
        already knows about the request before rendering starts — persisting
        it here is what lets get_status() report it back while still
        "generating", instead of only once the render finishes (see
        app/routers/satellite.py's acquire_lock() calls)."""
        json_path, lock_path = self._paths(key)
        json_path.unlink(missing_ok=True)
        lock_path.write_text(json.dumps(params or {}))

    def write_result(self, key: str, meta: dict) -> None:
        json_path, lock_path = self._paths(key)
        json_path.write_text(json.dumps(meta))
        lock_path.unlink(missing_ok=True)

    def output_path(self, key: str, suffix: str) -> Path:
        return self.base_dir / f"{key}.{suffix}"

    def list_keys(self) -> list[str]:
        """All keys with a status file (ready/error/generating-via-lock)."""
        json_keys = {p.stem for p in self.base_dir.glob("*.json")}
        lock_keys = {p.stem for p in self.base_dir.glob("*.lock")}
        return sorted(json_keys | lock_keys)

    def delete(self, key: str) -> int:
        """Remove every file for `key` (any suffix). Returns bytes freed."""
        freed = 0
        for p in self.base_dir.glob(f"{key}.*"):
            freed += p.stat().st_size
            p.unlink(missing_ok=True)
        return freed

    def stats(self) -> dict:
        count, total_bytes = 0, 0
        for p in self.base_dir.iterdir():
            if p.is_file():
                count += 1
                total_bytes += p.stat().st_size
        return {"file_count": count, "bytes": total_bytes}

File: crates/server/src/services/cache.rs
//! Disk-based result cache with a lock-file + TTL pattern — port of
//! `app/services/cache.py`. Backs the GOES tile pipeline: a `<key>.lock` marks
//! "rendering in progress", a `<key>.json` holds the final result.

use std::path::PathBuf;

use serde_json::{Map, Value};

pub struct ResultCache {
    base_dir: PathBuf,
    lock_timeout: u64,
}

impl ResultCache {
    pub fn new(base_dir: PathBuf, lock_timeout: u64) -> std::io::Result<Self> {
        std::fs::create_dir_all(&base_dir)?;
        Ok(Self { base_dir, lock_timeout })
    }

    fn paths(&self, key: &str) -> (PathBuf, PathBuf) {
        (
            self.base_dir.join(format!("{key}.json")),
            self.base_dir.join(format!("{key}.lock")),
        )
    }

    /// Port of `get_status`. Returns the ready/error result, a synthesized
    /// "generating" object (with elapsed seconds) while a fresh lock is held, or
    /// `None` if unknown / the lock went stale (which it also clears).
    pub fn get_status(&self, key: &str) -> Option<Value> {
        let (json_path, lock_path) = self.paths(key);
        if let Ok(text) = std::fs::read_to_string(&json_path) {
            if let Ok(meta) = serde_json::from_str::<Value>(&text) {
                if matches!(meta.get("status").and_then(Value::as_str), Some("ready" | "error")) {
                    return Some(meta);
                }
            }
        }
        if let Ok(md) = std::fs::metadata(&lock_path) {
            let age = md
                .modified()
                .ok()
                .and_then(|m| m.elapsed().ok())
                .map(|d| d.as_secs())
                .unwrap_or(0);
            if age > self.lock_timeout {
                let _ = std::fs::remove_file(&lock_path);
                return None;
            }
            // Start from whatever params acquire_lock persisted (band/cmap/etc.),
            // then overlay status/key/elapsed.
            let mut obj: Map<String, Value> = std::fs::read_to_string(&lock_path)
                .ok()
                .and_then(|t| serde_json::from_str::<Value>(&t).ok())
                .and_then(|v| v.as_object().cloned())
                .unwrap_or_default();
            obj.insert("status".into(), Value::from("generating"));
            obj.insert("key".into(), Value::from(key));
            obj.insert("elapsed".into(), Value::from(age));
            return Some(Value::Object(obj));
        }
        None
    }

    /// Port of `acquire_lock`: drop any prior result, write the params into the
    /// lock file so `get_status` can report them while still generating.
    pub fn acquire_lock(&self, key: &str, params: Option<&Value>) -> std::io::Result<()> {
        let (json_path, lock_path) = self.paths(key);
        let _ = std::fs::remove_file(&json_path);
        let body = params.cloned().unwrap_or_else(|| Value::Object(Map::new()));
        std::fs::write(&lock_path, serde_json::to_vec(&body).unwrap_or_default())
    }

    pub fn write_result(&self, key: &str, meta: &Value) -> std::io::Result<()> {
        let (json_path, lock_path) = self.paths(key);
        std::fs::write(&json_path, serde_json::to_vec(meta).unwrap_or_default())?;
        let _ = std::fs::remove_file(&lock_path);
        Ok(())
    }

    pub fn output_path(&self, key: &str, suffix: &str) -> PathBuf {
        self.base_dir.join(format!("{key}.{suffix}"))
    }

    /// All keys with a `.json` or `.lock` file.
    pub fn list_keys(&self) -> Vec<String> {
        let mut keys = std::collections::BTreeSet::new();
        if let Ok(entries) = std::fs::read_dir(&self.base_dir) {
            for entry in entries.flatten() {
                let path = entry.path();
                if matches!(path.extension().and_then(|e| e.to_str()), Some("json" | "lock")) {
                    if let Some(stem) = path.file_stem().and_then(|s| s.to_str()) {
                        keys.insert(stem.to_string());
                    }
                }
            }
        }
        keys.into_iter().collect()
    }

    /// Remove every file for `key`; returns bytes freed.
    pub fn delete(&self, key: &str) -> u64 {
        let mut freed = 0u64;
        let prefix = format!("{key}.");
        if let Ok(entries) = std::fs::read_dir(&self.base_dir) {
            for entry in entries.flatten() {
                let name = entry.file_name();
                if name.to_string_lossy().starts_with(&prefix) {
                    freed += entry.metadata().map(|m| m.len()).unwrap_or(0);
                    let _ = std::fs::remove_file(entry.path());
                }
            }
        }
        freed
    }

    pub fn stats(&self) -> Value {
        let (mut count, mut total) = (0u64, 0u64);
        if let Ok(entries) = std::fs::read_dir(&self.base_dir) {
            for entry in entries.flatten() {
                if entry.metadata().map(|m| m.is_file()).unwrap_or(false) {
                    count += 1;
                    total += entry.metadata().map(|m| m.len()).unwrap_or(0);
                }
            }
        }
        serde_json::json!({ "file_count": count, "bytes": total })
    }
}
